### PR TITLE
Optimize SNR time series in PyCBC Live

### PR DESCRIPTION
**Pull request by spxiwh**
We've been considering ways to improve the skymaps that are generated by PyCBC Live. One proposed method was, following the upload of an initial trigger, to go back and optimize the SNR time series of the trigger (mostly by finding the waveform with largest SNR). For e.g. high-mass triggers might be reported currently at lower masses because of higher glitchiness at high-masses. This adds a systematic bias. Here we propose a solution by uploading a second event after optimizing the SNR time series. This might take a little time to generate (O(minute)??) but can get an improved skymap quicker than the full PE, so is kind of an intermediate step.

 * This does increase the SNR, here's an example: https://gracedb-playground.ligo.org/events/T17156/view/ https://gracedb-playground.ligo.org/events/T17157/view/ where two triggers get uploaded for the same event.
 * I haven't tried adding a followup detector into the mix, the idea is that the coinc detectors would be optimized and then the followup added with the "new" template. That *should* work already.
 * What sort of sanity checks should be done after optimizing? Should we demand time- phase- etc. consistency after maximizing?
 * Optimizing is pretty slow with the current settings (1-2 minutes). I can reduce time at the cost of accuracy, so playing with this could be useful. I have this reduced to a play case here in CIT `/home/spxiwh/tmp_testoptimize`. Maybe I chose a bad test case.
 * I need to be able to disable the FFT context to enable the optimization to use single-threaded FFTs. (The changes to matched-filter to take this apart are obviously not intended to be merged). How can I run the optimize function with a single-core FFT context?
 * I added subsample interpolation. It does little in this case.
 * The f-lower is 30 Hz. Lowering that will gain a similar increase in SNR. We should do that here (if possible).
 * Should we spawn this into a new process to stop delaying the rest of the work while this is happening? What's the best way of doing this?
 * This should be proof against failures. E.g. if something doesn't work here, don't upload a second event and just keep going.

### CODE REVIEW

**Comment by spxiwh** [comment]
Okay, I still think this is important to do, but to avoid complicating `pycbc_live` my current plan is to get `pycbc_live` to dump the necessary information to file (data, PSDs, and some parameters) and then spawn a new job to do the optimization. This will allow `pycbc_live` to continue processing data, while this runs.

**Comment by ahnitz** [comment]
@spxiwh If you want to have a separate script that's fine, but you should spawn it off of pycbc live with mpi. See here for an example. 
https://mpi4py.readthedocs.io/en/stable/tutorial.html#dynamic-process-management

This also allows us to set a desired host for the new process (say an empty machine just for this purpose) and communicate some information if you want. You could write out to a file, or just message pass the data directly. The only reason we've avoided too much non-local file writing in past is that it has tended to be less reliable, but maybe that's no longer a concern on CIT.

**Comment by spxiwh** [comment]
@titodalcanton @ahnitz Okay, here's the reworked version of this patch. Now the changes to `pycbc_live` are reduced, so that `pycbc_live` will just dump the necessary information to file and then spawn a new process to run `optimize_snr` ... Alex suggested doing this in MPI, but given the relative time between saving/reading to disk and uploading to GraceDB, this step is negligible and so I would leave it for a future improvement.

In testing this I spotted two bugs in `pycbc_live`, which are fixed here:

 * The `singles_dtype` stored in coinc.py can be broken if when this is set the list of triggers is empty. In this case the dtype for `approximant` is not correctly recorded as a string, and the code will fail.
 * It is possible for there to be two uploads within a second (it's a weird edge-case, and it's basically the *same* trigger becoming significant just before the edge of an analysis boundary). The code fails in this place trying to duplicate the HDF file (this is a failure case before this patch). This can be resolved by adding a random string to the file names to allow for multiple triggers in the same second.
 * I also added the `analyst_comment` tag to comments so that they appear more clearly on the GraceDB pages.

Then this has been run on Tito's injection example set. You can see a bunch of the output of this here:

https://gracedb-playground.ligo.org/latest/?query=pycbc+test&query_type=E&get_neighbors=&results_format=

The `optimize_snr` job is only allowed (currently) to run for 150s before being killed, so some jobs currently do not upload a second trigger, but this does not affect the `pycbc_live` process. This has run stably on the injections provided by Tito.

I think I might still want to tweat the time-out limits ... and BNS could/should be made more efficient. But this is, I think, sufficient to go into the replay test analysis.

**Comment by spxiwh** [comment]
I've just made a few updates to this (now the followup ifos are always included in the network SNR number and all of the graceDB fields are populated, as with normal uploads).

I'm ready for this to go towards production now! Are there other changes/tests etc. that you want me to do?

File: bin/pycbc_optimize_snr [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+kwargs = {'psds': {ifo: followup_data[ifo]['psd'] for ifo in ifos},
+          'low_frequency_cutoff': flow,
+          'followup_data': followup_data,
+          'channel_names': channel_names}
+comments = ['Automatic PyCBC followup to find largest SNR parameters']
+
+doc = live.SingleCoincForGraceDB(coinc_ifos, coinc_results,
+                                 upload_snr_series=True,
+                                 followup_ifos=fup_ifos, **kwargs)
+gracedb_server = 'https://gracedb-playground.ligo.org/api/'
```
Comment (titodalcanton) [inline bin/pycbc_optimize_snr]: @spxiwh did you intend this functionality to actually upload to GraceDB playground, or was this just test code?

Comment (titodalcanton) [reply bin/pycbc_optimize_snr]: I think even if this is intended, I recommend having an actual command line option for this, with a default to gracedb-playground (see the `pycbc_live` args).

Comment (spxiwh) [reply bin/pycbc_optimize_snr]: Yes, definitely, this should be a command line. In production this should upload to the production server.

**Comment by titodalcanton** [comment]
@spxiwh I just pushed a couple fixes, including one for the comment I made earlier. I have not looked hyper-carefully at all the diff yet, and there are still some bits I would do differently, but I will get this going on the replay analysis now.

**Comment by titodalcanton** [comment]
@spxiwh could you rebase this on master by the way? So we get autogating together with it.

**Comment by titodalcanton** [comment]
As a further request, I would like all overwhitened strains, PSDs and attributes to be saved into a single HDF file, and it may be desirable to upload that file to GraceDB as well.

**Comment by titodalcanton** [comment]
During one of the optimizations, I got
```
2019-07-31 16:13:07,383 Starting optimize SNR
2019-07-31 16:13:07,419 Following up GID G94166
2019-07-31 16:13:07,421 Starting optimization
/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py:454: UserWarning: differential_evolution: the 'workers' keyword has overridden updating='immediate' to updating='deferred'
  " updating='deferred'", UserWarning)
Traceback (most recent call last):
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/bin/pycbc_optimize_snr", line 168, in <module>
    args=extra_args)
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 272, in differential_evolution
    ret = solver.solve()
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 674, in solve
    self.population)
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 775, in _calculate_population_energies
    parameters_pop[0:nfevs]))
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/_lib/_util.py", line 414, in __call__
    return self._mapfunc(func, iterable)
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 250, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 554, in get
    raise self._value
AttributeError: 'FrequencySeries' object has no attribute 'params'
```

**Comment by spxiwh** [comment]
@titodalcanton I've just rebased this, which I think is the most critical thing to have this being tested along with autogating. I'll try and look at the comments when I find a minute.

**Comment by spxiwh** [comment]
Can I check. In my testing I've always been using `IMRPhenomD` as the approximant. I think the code will fail if you try `SPATmplt` ... which I can try to fix. (The htilde object needs a `params` attribute added, which stores the mass/spin/flow parameters, not 100% sure in what format).

**Comment by titodalcanton** [comment]
Ah right, that is probably the cause of the error. The first trigger had a heavy template, but the second one (which failed) had definitely a SPATmplt template. Why does SPATmplt not carry the params attribute though?

(BTW I just did a couple of small commits)

**Comment by spxiwh** [comment]
It's not that it carries the `params` attribute and the other does not. It's that `SPATmplt` is the only approximant that *needs* the `params` attribute when doing the sigma calculation. Most waveform approximants just use the `h(t)` for this, but `SPATmplt` uses an analytical approximant, that needs to know what the masses/spins are.

**Comment by titodalcanton** [comment]
Hmm. Different error now:
```
2019-08-01 07:47:43,864 Starting optimize SNR
2019-08-01 07:47:43,902 Following up GID G94249
2019-08-01 07:47:43,904 Starting optimization
/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py:454: UserWarning: differential_evolution: the 'workers' keyword has overridden updating='immediate' to updating='deferred'
  " updating='deferred'", UserWarning)
Traceback (most recent call last):
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/bin/pycbc_optimize_snr", line 170, in <module>
    args=extra_args)
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 272, in differential_evolution
    ret = solver.solve()
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 674, in solve
    self.population)
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 775, in _calculate_population_energies
    parameters_pop[0:nfevs]))
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/_lib/_util.py", line 414, in __call__
    return self._mapfunc(func, iterable)
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 250, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 554, in get
    raise self._value
AttributeError: 'FrequencySeries' object has no attribute 'end_idx'
```

**Comment by spxiwh** [comment]
I think its the same problem though. The `sigma` calculation needs some extra stuff. Looks like it needs:

https://github.com/gwastro/pycbc/blob/master/pycbc/waveform/bank.py#L593

for the termination frequency in the analytical calculation

**Comment by titodalcanton** [comment]
Nice, it is optimizing now.

**Comment by titodalcanton** [comment]
New error after a few more days. Sounds like we need more aggressive limiting of the parameters.
```
2019-08-04 04:34:22,485 Starting optimize SNR
2019-08-04 04:34:22,500 Following up GID G94578
2019-08-04 04:34:22,502 Starting optimization
XLAL Error - XLALSimIMRSEOBNRv4ROMTimeOfFrequency (LALSimIMRSEOBNRv4ROM.c:1448): Frequency 24.9 is outside allowed frequency range.
XLAL Error - XLALSimIMRSEOBNRv4ROMTimeOfFrequency (LALSimIMRSEOBNRv4ROM.c:1448): Input domain error
Traceback (most recent call last):
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/bin/pycbc_optimize_snr", line 172, in <module>
    args=extra_args)
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 272, in differential_evolution
    ret = solver.solve()
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 674, in solve
    self.population)
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/optimize/_differentialevolution.py", line 775, in _calculate_population_energies
    parameters_pop[0:nfevs]))
  File "/home/pycbc.live/testing/O2_replay/pycbc_live_O2_replay_env/lib/python2.7/site-packages/scipy/_lib/_util.py", line 414, in __call__
    return self._mapfunc(func, iterable)
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 250, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 554, in get
    raise self._value
RuntimeError: Input domain error
```

**Comment by titodalcanton** [comment]
@spxiwh please rebase on master when you have a chance, to include #2889.

**Comment by spxiwh** [comment]
This is rebased again

**Comment by titodalcanton** [comment]
@spxiwh I think I fixed the problem with approximant failures. This is getting very close to a seal of approval, my last request being the single HDF file containing all the results. I will work on this unless you get to it first.

**Comment by titodalcanton** [comment]
Some more tweaking is required. BAYESTAR cannot run on the optimized upload:
```
ERROR:BAYESTAR:sky localization failed
Traceback (most recent call last):
  File "/home/emfollow-playground/gwcelery/tasks/bayestar.py", line 76, in localize
    skymap = _bayestar.localize(event)
  File "/home/emfollow-playground/.local/lib/python3.6/site-packages/ligo/skymap/bayestar/__init__.py", line 347, in localize
    f_high_truncate=f_high_truncate)
  File "/home/emfollow-playground/.local/lib/python3.6/site-packages/ligo/skymap/bayestar/__init__.py", line 236, in condition
    'but {} was off by {} s'.format(ifo, shift))
ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but H1 was off by 0.006591798249999999 s
```

**Comment by spxiwh** [comment]
@titodalcanton Do you have an example of an upload failing to produce a skymap? I am seeing Bayestar skymaps being produced on the testing data.

**Comment by titodalcanton** [comment]
@spxiwh here is an example from [a recent optimized upload](https://gracedb-playground.ligo.org/events/G122225/view/):
```
$ bayestar-localize-coincs coinc.xml coinc.xml
2019-09-20 09:54:43,582 INFO Using 4 OpenMP thread(s)
2019-09-20 09:54:43,583 INFO coinc.xml,coinc.xml:reading input files
2019-09-20 09:54:43,868 WARNING Time slide record is missing for time_slide:time_slide_id:0, guessing that this is zero-lag
2019-09-20 09:54:43,869 INFO coinc_event:coinc_event_id:0:computing sky map
2019-09-20 09:54:44,088 WARNING Truncating PSD at 0.95 of maximum frequency to suppress rolloff artifacts. This option may be removed in the future.
2019-09-20 09:54:44,091 WARNING Truncating PSD at 0.95 of maximum frequency to suppress rolloff artifacts. This option may be removed in the future.
2019-09-20 09:54:44,093 WARNING Truncating PSD at 0.95 of maximum frequency to suppress rolloff artifacts. This option may be removed in the future.
2019-09-20 09:54:44,095 WARNING Template is unspecified; using ER10/O2 uberbank criterion
2019-09-20 09:54:44,095 INFO Selected template: SEOBNRv4_ROM
2019-09-20 09:54:44,291 WARNING Assuming PSD is infinite at 0.03125 Hz because PSD is only sampled down to 19.8 Hz
2019-09-20 09:54:44,291 WARNING Assuming PSD is infinite at 2047.97 Hz because PSD is only sampled up to 972.8 Hz
2019-09-20 09:54:44,297 WARNING Assuming PSD is infinite at 0.03125 Hz because PSD is only sampled down to 19.8 Hz
2019-09-20 09:54:44,297 WARNING Assuming PSD is infinite at 2047.97 Hz because PSD is only sampled up to 972.8 Hz
2019-09-20 09:54:44,301 WARNING Assuming PSD is infinite at 0.03125 Hz because PSD is only sampled down to 19.8 Hz
2019-09-20 09:54:44,302 WARNING Assuming PSD is infinite at 2047.97 Hz because PSD is only sampled up to 972.8 Hz
2019-09-20 09:54:44,369 ERROR coinc_event:coinc_event_id:0:sky localization failed
Traceback (most recent call last):
  File "/home/altivec/downloads/py37_env/lib/python3.7/site-packages/ligo/skymap/tool/bayestar_localize_coincs.py", line 164, in main
    f_high_truncate=opts.f_high_truncate)
  File "/home/altivec/downloads/py37_env/lib/python3.7/site-packages/ligo/skymap/bayestar/__init__.py", line 347, in localize
    f_high_truncate=f_high_truncate)
  File "/home/altivec/downloads/py37_env/lib/python3.7/site-packages/ligo/skymap/bayestar/__init__.py", line 236, in condition
    'but {} was off by {} s'.format(ifo, shift))
ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but H1 was off by 0.046142576500000004 s
Traceback (most recent call last):
  File "/home/altivec/downloads/py37_env/bin/bayestar-localize-coincs", line 10, in <module>
    sys.exit(main())
  File "/home/altivec/downloads/py37_env/lib/python3.7/site-packages/ligo/skymap/tool/bayestar_localize_coincs.py", line 164, in main
    f_high_truncate=opts.f_high_truncate)
  File "/home/altivec/downloads/py37_env/lib/python3.7/site-packages/ligo/skymap/bayestar/__init__.py", line 347, in localize
    f_high_truncate=f_high_truncate)
  File "/home/altivec/downloads/py37_env/lib/python3.7/site-packages/ligo/skymap/bayestar/__init__.py", line 236, in condition
    'but {} was off by {} s'.format(ifo, shift))
ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but H1 was off by 0.046142576500000004 s
```
I will probably not get to fixing this soon, but the error seems pretty informative.

**Comment by spxiwh** [comment]
I figured out where the problem is coming from. I've added a fix, but still need to test it.

**Comment by lpsinger** [comment]
This is doing inter-template interpolation, basically?

**Comment by spxiwh** [comment]
@lpsinger Yes ... I'd probably say that this is trying to maximize the SNR over the mass/spin space, but I guess that's the same thing.

**Comment by titodalcanton** [comment]
@spxiwh I think the chi^2 = 1 hack is no longer necessary (since [this](https://git.ligo.org/emfollow/gwcelery/merge_requests/644) should fix the issue "cleanly"). I am having a hard time judging whether the whole infrastructure is supposed to work on playground at the moment though (no labels are being applied anymore, even with the chi^2 = 1 hack). Given that this new functionality is enabled/disabled with a clear command-line switch, I propose reverting the chi^2 = 1 hack and merging this now. We will then continue addressing any remaining issues (e.g. putting back the chi^2 = 1 hack if still necessary) as further merge requests.

**Comment by spxiwh** [comment]
@titodalcanton Okay. If this (at the current version) has been successfully uploading events to playground (optimized and un-optimized) then I'm happy to merge. We should check with the EM followup folks that everything is working as intended downstream.

**Comment by titodalcanton** [comment]
Unfortunately it looks like the BAYESTAR issue is still there. I just tried localizing one of the latest uploads and I still get
```
ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but V1 was off by 0.024658202249999983 s
```

**Comment by spxiwh** [comment]
@titodalcanton Do you have the inputs and command used to run the optimize SNR code? I could debug this more easily if I have a clear example to test on.

**Comment by lpsinger** [comment]
> Unfortunately it looks like the BAYESTAR issue is still there. I just tried localizing one of the latest uploads and I still get
> 
> ```
> ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but V1 was off by 0.024658202249999983 s
> ```

This error message looks like the timestamp of the SNR series might be incorrect. The end_time/end_time_ns fields in the sngl_inspiral may be interpolated if you like, but the time stamp of the SNR series should be the time stamp *of its first sample*. You might be seeing this error message if you accidentally set the SNR series time stamp to the interpolated time.

**Comment by titodalcanton** [comment]
@spxiwh it seems to fail on all uploads, but here is an example with specific commands:
```
pycbc_optimize_snr --data-files V1:/local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_V1_data_overwhitened.hdf H1:/local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_H1_data_overwhitened.hdf L1:/local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_L1_data_overwhitened.hdf --psd-files V1:/local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_V1_psd.hdf H1:/local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_H1_psd.hdf L1:/local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_L1_psd.hdf --params-file /local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_attributes.hdf --approximant SEOBNRv4_ROM --gracedb-server https://gracedb-playground.ligo.org/api/ --production --verbose &> /local/pycbc.live/O2_replay/triggers/coinc-1256242531-cveezp_optimize_snr_log.dat
```
(note that the above files in /local will soon be moved to the path in ~). I downloaded the `coinc.xml` file from GraceDB, then
```
$ bayestar-localize-coincs coinc.xml coinc.xml 
2019-10-30 03:47:39,547 INFO Using 1 OpenMP thread(s)
2019-10-30 03:47:39,548 INFO coinc.xml,coinc.xml:reading input files
2019-10-30 03:47:39,737 WARNING Time slide record is missing for time_slide:time_slide_id:0, guessing that this is zero-lag
2019-10-30 03:47:39,738 INFO coinc_event:coinc_event_id:0:computing sky map
2019-10-30 03:47:39,905 WARNING Truncating PSD at 0.95 of maximum frequency to suppress rolloff artifacts. This option may be removed in the future.
2019-10-30 03:47:39,907 WARNING Truncating PSD at 0.95 of maximum frequency to suppress rolloff artifacts. This option may be removed in the future.
2019-10-30 03:47:39,909 WARNING Truncating PSD at 0.95 of maximum frequency to suppress rolloff artifacts. This option may be removed in the future.
2019-10-30 03:47:39,910 WARNING Template is unspecified; using ER10/O2 uberbank criterion
2019-10-30 03:47:39,910 INFO Selected template: SEOBNRv4_ROM
2019-10-30 03:47:40,392 WARNING Assuming PSD is infinite at 0.0625 Hz because PSD is only sampled down to 19.6 Hz
2019-10-30 03:47:40,392 WARNING Assuming PSD is infinite at 2047.94 Hz because PSD is only sampled up to 972.85 Hz
2019-10-30 03:47:40,394 WARNING Assuming PSD is infinite at 0.0625 Hz because PSD is only sampled down to 19.6 Hz
2019-10-30 03:47:40,394 WARNING Assuming PSD is infinite at 2047.94 Hz because PSD is only sampled up to 972.85 Hz
2019-10-30 03:47:40,397 WARNING Assuming PSD is infinite at 0.0625 Hz because PSD is only sampled down to 19.6 Hz
2019-10-30 03:47:40,397 WARNING Assuming PSD is infinite at 2047.94 Hz because PSD is only sampled up to 972.85 Hz
2019-10-30 03:47:40,402 ERROR coinc_event:coinc_event_id:0:sky localization failed
Traceback (most recent call last):
  File "/home/tito/virtualenvs/ligo.skymap_test/lib/python3.6/site-packages/ligo/skymap/tool/bayestar_localize_coincs.py", line 164, in main
    f_high_truncate=opts.f_high_truncate)
  File "/home/tito/virtualenvs/ligo.skymap_test/lib/python3.6/site-packages/ligo/skymap/bayestar/__init__.py", line 347, in localize
    f_high_truncate=f_high_truncate)
  File "/home/tito/virtualenvs/ligo.skymap_test/lib/python3.6/site-packages/ligo/skymap/bayestar/__init__.py", line 236, in condition
    'but {} was off by {} s'.format(ifo, shift))
ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but H1 was off by 0.019775389749999997 s
Traceback (most recent call last):
  File "/home/tito/virtualenvs/ligo.skymap_test/bin/bayestar-localize-coincs", line 8, in <module>
    sys.exit(main())
  File "/home/tito/virtualenvs/ligo.skymap_test/lib/python3.6/site-packages/ligo/skymap/tool/bayestar_localize_coincs.py", line 164, in main
    f_high_truncate=opts.f_high_truncate)
  File "/home/tito/virtualenvs/ligo.skymap_test/lib/python3.6/site-packages/ligo/skymap/bayestar/__init__.py", line 347, in localize
    f_high_truncate=f_high_truncate)
  File "/home/tito/virtualenvs/ligo.skymap_test/lib/python3.6/site-packages/ligo/skymap/bayestar/__init__.py", line 236, in condition
    'but {} was off by {} s'.format(ifo, shift))
ValueError: BAYESTAR expects the SNR time series to be centered on the single-detector trigger times, but H1 was off by 0.019775389749999997 s
```

**Comment by spxiwh** [comment]
@titodalcanton Thanks for the example (and thanks @lpsinger for the suggestion). The problem here was me not really understanding how the end_time is set for PyCBC Live's "followup" ifos. I was setting the end time correctly in the updated patch (I think), but it was being overridden in `io/live.py`. This should *now* be correct and not being overridden, and I managed to get a "SKYMAP READY" for this example:

https://gracedb-playground.ligo.org/events/G179920/view/

.... As a note this seems to be an issue with how preferred events are selected. We *have* to include all ifo's SNRs in the sngl_inspiral table because even missing an SNR of 3 will mean that PyCBC would never be considered. However, we really don't know the event's end time particularly well in such a quiet detector, so quoting a single time is strange. This would beg the question that if we have 2 "quiet" ifos how would we pick times so that all time delays are physical ... Or *should* we do that at all ... I think I'm rambling at this point, the patch should hopefully work now.

**Comment by lpsinger** [comment]
> However, we really don't know the event's end time particularly well in such a quiet detector, so quoting a single time is strange.

You're fine as long as your sub threshold detectors' SNR time series covers all plausible lags relative to the detectors that triggered.

You can leave the time in the sngl_inspiral blank, if you want. It's only used for sanity-checking the SNR time series offset.

**Comment by titodalcanton** [comment]
Thanks @spxiwh, I am now going to merge this and get it running on the replay analysis.

**Review by titodalcanton — APPROVED** [review]

**Pull request merged**